lib/resty/core/socket.lua

```diff
@@ -0,0 +1,85 @@
+local base = require "resty.core.base"
+local debug = require 'debug'
+local ffi = require 'ffi'
+
+
+local error = error
+local tonumber = tonumber
+local registry = debug.getregistry()
+local ffi_new = ffi.new
+local ffi_string = ffi.string
+local C = ffi.C
+
+
+ffi.cdef[[
+typedef struct ngx_http_lua_socket_tcp_upstream_s
+    ngx_http_lua_socket_tcp_upstream_t;
+
+int 
+ngx_http_lua_ffi_socket_tcp_getoption(ngx_http_lua_socket_tcp_upstream_t *u,
+    const char *opt, int *val, char **err_msg);
+
+int 
+ngx_http_lua_ffi_socket_tcp_setoption(ngx_http_lua_socket_tcp_upstream_t *u,
+    const char *opt, int val, char **err_msg); 
+]]
+
+local output_value_buf = ffi_new("int[1]")
+local err_msg = base.get_errmsg_ptr()
+local FFI_OK = base.FFI_OK
+local SOCKET_CTX_INDEX = 1
+
+
+local function get_tcp_socket(cosocket)
+    local tcp_socket = cosocket[SOCKET_CTX_INDEX]
+    if not tcp_socket then
+        return nil, error("bad tcp socket")
+    end
+
+    return tcp_socket
+end
+
+local function getoption(cosocket, opt)
+    local tcp_socket = get_tcp_socket(cosocket)
+
+    if opt == nil then 
+        return nil, "nil option"
+    end
+
+    local rc = C.ngx_http_lua_ffi_socket_tcp_getoption(tcp_socket,
+                                                       opt, 
+                                                       output_value_buf,
+                                                       err_msg)
+    if rc ~= FFI_OK then 
+        return nil, ffi_string(err_msg[0])
+    end
+
+    return tonumber(output_value_buf[0])
+end
+
+
+local function setoption(cosocket, opt, val)
+    local tcp_socket = get_tcp_socket(cosocket)
+
+    if opt == nil or val == nil then 
+        return nil, "nil option"
+    end
+
+    local rc = C.ngx_http_lua_ffi_socket_tcp_setoption(tcp_socket, 
+                                                       opt,
+                                                       val,
+                                                       err_msg)
+    if rc ~= FFI_OK then 
+        return nil, ffi_string(err_msg[0])
+    end
+
+    return true
+end
+
+do
+    local method_table = registry.__tcp_cosocket_mt
+    method_table.getoption = getoption
+    method_table.setoption = setoption
+end
+
+return { version = base.version }
```

Comment:
we usually keep the `\n` for the last line.

Reply:
accept